didicloud/Classes/Storage.swift

```diff
@@ -235,4 +235,123 @@ public struct Storage {
+                let dispatchGroup = DispatchGroup()
+                
+                for value in values {
+                    
+                    dispatchGroup.enter()
+                    
+                    guard let recordID = value.recordID else {
+                        completion(.failure(StorageError.cloudKitNullReference))
+                        return
+                    }
+                    
+                    remove(storageType: storageType, recordID) {
+                        (result: Result<CKRecord.ID, Error>) in
+                        
+                        switch result {
+                            
+                        case .failure(_):
+                            completion(.failure(StorageError.cloudKitDataRemoval))
+                            return
+                        
+                        case .success(let recordID):
+                            deletedIDs.append(recordID)
+                            dispatchGroup.leave()
+                            
+                        }
+                    }
+                }
+                
+                dispatchGroup.notify(queue: .main) {
+                    completion(.success(deletedIDs))
+                }
+            }
+        }
+    }
+    
+    /// Removes all records of type T owned by current user
+    /// - Parameters:
+    ///   - storageType: Which database to perform the query
+    ///   - type: The type of the record
+    ///   - completion: Result object the deleted record ID's or an error
+    public static func removeAllbyUser<T: Storable>(storageType: StorageType = .privateStorage, type: T.Type, completion: @escaping (Result<[CKRecord.ID], Error>) -> Void) {
+        
+        getUserRecordID { (result) in
+            
+            switch result {
+           
+            case .success(let recordID):
+               
+                let query = CKQuery(
+                    recordType: T.reference,
+                    predicate: NSPredicate(format: "creatorUserRecordID == %@", recordID)
+                )
+                
+                storageType.database.perform(query, inZoneWith: nil) {
+                    results, error in
+                    
+                    if error != nil {
+                        completion(.failure(StorageError.cloudKitDataRetrieval))
+                        return
+                    }
+                    
+                    guard let results = results else {
+                        completion(.failure(StorageError.cloudKitNullReturn))
+                        return
+                    }
+                    
+                    var deletedIDs: [CKRecord.ID] = []
+                    let dispatchGroup = DispatchGroup()
+                    
+                    for value in results {
+                        
+                        dispatchGroup.enter()
+                        
+                        remove(storageType: storageType, value.recordID) {
```

Comment:
Não procurou se existe uma função do cloudkit para deletar varios records? Porque do jeito que tá não escala. Vai abrir uma conexão pra cada record que tu for deletar (N conexões no total). Pesquisa rápida no Google tu tem um deleteAllByRecordIDs: 

```
func deleteAllRecords() {
    let publicDatabase: CKDatabase = CKContainer.defaultContainer().publicCloudDatabase
    // fetch records from iCloud, get their recordID and then delete them
    var recordIDsArray: [CKRecordID] = []
    let operation = CKModifyRecordsOperation(recordsToSave: nil, recordIDsToDelete: recordIDsArray)
   operation.modifyRecordsCompletionBlock = {
       (savedRecords: [CKRecord]?, deletedRecordIDs: [CKRecordID]?, error: NSError?) in
        print("deleted all records")
    }
    publicDatabase.addOperation(operation)
}
```

Usando a fetchRecordsByUser() tu consegue o recordID de cada record, monta o array de recordIDs e ta feita a função. Isso usando 2 aberturas de conexão, não N

Reply:
Criei o método  `remove(_ recordIDs: [CKRedord.ID])` utilizando `CKModifyRecordsOperation`, e ajustei os métodos `removeAll` e `removeAllByUser` para o utilizarem. Agora é utilizada só uma abertura de conexão, e excluí a necessidade de usar o `DispatchGroup`.